compiler/internal/dce/info.go

```diff
@@ -0,0 +1,75 @@
+package dce
+
+import (
+	"go/types"
+	"sort"
+
+	"github.com/gopherjs/gopherjs/compiler/typesutil"
+)
+
+// Info contains information used by the dead-code elimination (DCE) logic to
+// determine whether a declaration is alive or dead.
+type Info struct {
+
+	// importPath is the package path of the package the declaration is in.
+	importPath string
+
+	// Symbol's identifier used by the dead-code elimination logic, not including
+	// package path. If empty, the symbol is assumed to be alive and will not be
+	// eliminated. For methods it is the same as its receiver type identifier.
+	objectFilter string
+
+	// The second part of the identified used by dead-code elimination for methods.
+	// Empty for other types of symbols.
+	methodFilter string
+
+	// List of fully qualified (including package path) DCE symbol identifiers the
+	// symbol depends on for dead code elimination purposes.
+	deps []string
+}
+
+// isAlive returns true if the declaration is marked as alive.
+func (d *Info) isAlive() bool {
+	return d.objectFilter == "" && d.methodFilter == ""
+}
+
+// SetAsAlive marks the declaration as alive, meaning it will not be eliminated.
+// This must be done after the SetName method is called.
```

Comment:
Let's make this method panic if this invariant is violated.

Reply:
Instead of a panic, I made the explicitly marking as alive use an `alive` flag instead of clearing out the names. Now, if the names aren't set or if the `alive` flag is set true, the decl is alive. This allows the decl to be marked as alive before or after naming without effecting the naming.